app/models/annotation/AnnotationUploadService.scala

```diff
@@ -15,48 +15,65 @@ import net.liftweb.common.{Box, Empty, Failure, Full}
 import net.liftweb.common.Box.tryo
 import play.api.i18n.MessagesProvider
 
+import scala.concurrent.ExecutionContext
+
 case class UploadedVolumeLayer(tracing: VolumeTracing, dataZipLocation: String, name: Option[String]) {
   def getDataZipFrom(otherFiles: Map[String, File]): Option[File] =
     otherFiles.get(dataZipLocation)
 }
 
 class AnnotationUploadService @Inject()(tempFileService: TempFileService) extends LazyLogging {
 
-  private def extractFromNml(file: File, name: String, overwritingDataSetName: Option[String], isTaskUpload: Boolean)(
-      implicit m: MessagesProvider): NmlParseResult =
-    extractFromNml(new FileInputStream(file), name, overwritingDataSetName, isTaskUpload)
+  private def extractFromNmlFile(
+      file: File,
+      name: String,
+      overwritingDataSetName: Option[String],
+      userToken: Option[String],
+      isTaskUpload: Boolean)(implicit m: MessagesProvider, ec: ExecutionContext): NmlParseResult =
+    extractFromNml(new FileInputStream(file), name, overwritingDataSetName, isTaskUpload, userToken)
 
   private def formatChain(chain: Box[Failure]): String = chain match {
     case Full(failure) =>
       " <~ " + failure.msg + formatChain(failure.chain)
     case _ => ""
   }
 
-  def extractFromNml(inputStream: InputStream,
-                     name: String,
-                     overwritingDataSetName: Option[String],
-                     isTaskUpload: Boolean,
-                     basePath: Option[String] = None)(implicit m: MessagesProvider): NmlParseResult =
-    NmlParser.parse(name, inputStream, overwritingDataSetName, isTaskUpload, basePath) match {
+  def extractFromNml(
+      inputStream: InputStream,
+      name: String,
+      overwritingDataSetName: Option[String],
+      isTaskUpload: Boolean,
+      userToken: Option[String],
+      basePath: Option[String] = None)(implicit m: MessagesProvider, ec: ExecutionContext): NmlParseResult =
+    NmlParser
+      .parse(name, inputStream, overwritingDataSetName, isTaskUpload, basePath, None, userToken)
+      .await("NMLParser/parse was changed to return Fox in #7437. Removing this await is tracked in #7551") match {
```

Comment:
What is the timeout of the await? Parsing large NML files can take some time. Maybe it would be possible to move the await into the parser so that only the thing that really has to be async (looking up the segment index file existence) is awaited?

Reply:
Moved it; the timeout is 5 seconds currently